modules/standard/Random.chpl

```diff
@@ -222,14 +291,332 @@ module Random {
     randNums.permutation(arr);
   }
 
+  /*
+    Use a new :record:`randomStream` to produce a random permutation of an
+    array's domain. The values ``d.dim(0).low..d.dim(0).high`` will appear
+    exactly once in the array in a pseudo-random order.
+
+    :arg arr: The array to store the permutation in
+    :arg seed: The seed to use when initializing a ``randomStream``
+  */
+  @unstable("'permutation' is unstable and subject to change")
+  proc permutation(ref arr: [?d] ?t, seed: int)
+    where isCoercible(d.idxType, t) && is1DRectangularDomain(d) do
+  {
+    var rs = new randomStream(d.idxType, seed);
+    rs.permutation(arr);
+  }
+
+  /*
+    Use a new :record:`randomStream` to produce a random permutation of an
+    array's domain. The values ``d.dim(0).low..d.dim(0).high`` will appear
+    exactly once in the array in a pseudo-random order.
+
+    :arg arr: The array to store the permutation in
+  */
+  @unstable("'permutation' is unstable and subject to change")
+  proc permutation(ref arr: [?d] ?t)
+    where isCoercible(d.idxType, t) && is1DRectangularDomain(d) do
+  {
+    var rs = new randomStream(d.idxType);
+    rs.permutation(arr);
+  }
+
+  @chpldoc.nodoc
+  proc permutation(ref arr: [], seed: int) {
+    compilerError("'permutation' only supports 1D rectangular arrays whose domain idxType is compatible with eltType");
+  }
+
   @chpldoc.nodoc
-  proc makeRandomStream(type eltType,
-                        seed: int(64) = SeedGenerator.oddCurrentTime,
-                        param parSafe: bool = true,
-                        param algorithm = defaultRNG) {
-    compilerWarning("makeRandomStream is deprecated - " +
-                    "please use createRandomStream instead");
-    return createRandomStream(eltType, seed, parSafe, algorithm);
+  proc permutation(ref arr: []) {
+        compilerError("'permutation' only supports 1D rectangular arrays whose domain idxType is compatible with eltType");
+  }
+
+  record randomStream: writeSerializable {
+    /*
+      Specifies the type of value generated by the random stream. Currently
+      numeric and bool types are supported.
+    */
+    type eltType;
+
+    /* The seed value for the PCG random number generator */
+    const seed: int;
+
+    @chpldoc.nodoc
+    var pcg: shared PCGRandomStreamInternal(eltType, false);
+
+    /*
+      Create a new ``randomStream`` using the specified seed and parallel
+      safety.
+    */
+    proc init(type eltType, seed: int) where isNumericOrBoolType(eltType) {
+      this.eltType = eltType;
+      this.seed = seed;
+      this.pcg = new shared PCGRandomStreamInternal(eltType, seed, false);
+    }
+
+    /*
+      Create a new ``randomStream`` using the specified parallel safety and
+      a seed computed from the current time.
```

Comment:
Let's say the seed is computed in an implementation-dependent manner that depends upon the current time.